stream/src/main/scala/org/apache/pekko/stream/stage/GraphStage.scala

```diff
@@ -1282,7 +1283,9 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
   private var callbacksWaitingForInterpreter: List[ConcurrentAsyncCallback[_]] = Nil
   // is used for two purposes: keep track of running callbacks and signal that the
   // stage has stopped to fail incoming async callback invocations by being set to null
-  private val asyncCallbacksInProgress = new AtomicReference[List[Promise[Done]]](Nil)
+  // Using ConcurrentHashMap's KeySetView as Set to track the inProgress async callbacks.
+  private val asyncCallbacksInProgress: AtomicReference[java.util.Set[Promise[Done]]] =
+    new AtomicReference(new ConcurrentHashMap[Promise[Done], NotUsed]().keySet(NotUsed))
```

Comment:
why use an AtomicReference on a mutable data structure like a ConcurrentHashMap? why not just mutate the map?

Reply:
The `AtomicRef` is needed as the callback can be called in a `AsyncCallback` which can be accessed from another thread.
Otherwise it will be a `@volatile asyncCallbacksInProgress` I think to indicate the stage has been stoped.

Reply:
@pjfanning do you suggest I chang it to `var @volatile asyncCallbacksInProgress` ?

Reply:
ConcurrentHashMaps are thread safe. I would suggest that we should look at making asyncCallbacksInProgress a val and we can do the usual Map calls on the Map - get, put, clear, etc.

Reply:
That will not work, because an empty `ConcurrnetHashMap` can't indicate whether the stage is stopped. you can take a look at the comments upper 2~3 lines of this line.

Reply:
> That will not work, because an empty `ConcurrnetHashMap` can't indicate whether the stage is stopped. you can take a look at the comments upper 2~3 lines of this line.

Keep the AtomicReference then but could could to create the Set using this?

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#newKeySet--

Reply:
But the default value is BOOLEAN.TRUE, seems a little weird in this case.

Reply:
Does it matter what the value type is? The Map needs a value type but you are only interested in the key type.

Reply:
This saves use from having to create a map and then create a keyset - there is just one call to create a keyset.

Reply:
Noop, it will still create a map, I will update the pr.

Reply:
@pjfanning I have updated as you suggestion.